### PR TITLE
Query Editor: Support selecting multiple property aliases in time series queries

### PR DESCRIPTION
**Pull request by hwandersman**
**What this PR does / why we need it**:

The most pressing feature customers are asking for is to support multi-value template variables and manually selecting multiple propertyIds / property aliases in time series queries. 

This PR enables users to do the following:
1. **Enter multiple property aliases** for GetPropertyValue, GetPropertyValueHistory, GetInterpolatedPropertyValues, and GetPropertyValueAggregates queries
1. **Enter/select multiple template variables** (including multi-value variables) for the **property alias** field of GetPropertyValue, GetPropertyValueHistory, GetInterpolatedPropertyValues, and GetPropertyValueAggregates queries
1. **Enter/select single template variable** ~~(including multi-value variables)~~ for the **propertyId** field of GetPropertyValue, GetPropertyValueHistory, GetInterpolatedPropertyValues, and GetPropertyValueAggregates queries
  * Update (25/03/06): single property id selection in UI until a better UX is designed around pairing assets and properties

**Considerations**

* Backwards compatibility: anyone who upgrades the plugin with this change will not have broken dashboards
* API batching: each batch is made up of either a list of pairs of a assetIds and propertyIds or a list of property aliases
* API batch limits and handling nextTokens between batches: already considered when supporting multiple assetIds, now the entries are based on all id and alias inputs

**Missing edge cases**

I've identified the following edge cases that are NOT handled in this PR (and are pre-existing - not introduced in this PR):
* Handling [skipped entries](https://docs.aws.amazon.com/iot-sitewise/latest/APIReference/API_BatchGetAssetPropertyValueHistory.html#iotsitewise-BatchGetAssetPropertyValueHistory-response-skippedEntries) in the batch API responses. Entries that have already completed but need to remain in the query is not processed in further frames, and leaving it out breaks subsequent calls with nextToken.
* User can select two assets from different asset models with inconsistent results
* User cannot add assets via externalId in the list with selected inputs. Can either have a list of assets via externalId or selected assets
* When user selects an aggregate type (Average, Min, Max, etc.) that is not supported by a selected property we silently do not query it. We should display some error especially now with multiple properties where some have support and some may not.
* Query options appearing without content when property aliases are set for GetPropertyValue query

Other edge cases to test for:
* Edge APIs - no batching still works
* L4E formatting

**Design decisions**

Flow for time series API handlers:
1. Fetch all associated assetIds and propertyIds - relevant for property alias arguments. Add content to an *AssetPropertyEntry* list which contains each data stream's assetId, propertyId, and property alias.
3. [For all queries but GetInterpolatedPropertyValues]: Split the query into batches based on the API max entry limit, where each batch has a list of *AssetPropertyEntries* in a separate API call
4. Build the API request for each *AssetPropertyEntry* defaulted to an assetId and propertyId, or a property alias if the data stream is disassociated
5. Submit the API request and build data frames with the responses
6. In the frontend of the datasource assign a *NextTokens* field mapped by the *entryId*. This is used in subsequent API requests with the same entryId, built by a hash of the *assetId-propertyId* or property alias.

**Test coverage**

* Backend unit tests for all time series queries handling multiple assetIds, propertyIds, and property aliases
* Frontend unit tests updated to use lists for assetIds, propertyIds, and property aliases
* TODO: add E2E test coverage for multi-selection

**Manual tests**

1. Get property value history with 4 assetIds and 5 propertyIds (20 entries > 16 API entry limit)

image

3. Get property value aggregates with 4 assetIds and 5 propertyIds (20 entries > 16 API entry limit)

image

4. Get property value with 4 assetIds and 16 propertyIds (64 entries < 128 API entry limit)

image

5. Get interpolated property values with 1 assetId and 3 propertyIds and a resolution of 10 seconds (to ensure there are nextTokens). This takes seconds to load because we set max results per request to 10 which seems very low.

image

6. Get property value history with 48 property aliases set by a multi-value template variable (48 entries > 16 API entry limit)

image

7. Get property value aggregates with 48 property aliases set by a multi-value template variable (47 entries > 16 API entry limit)

image

8. Get property value with 48 property aliases set by a multi-value template variable (48 entries < 128 API entry limit)

image

9. Get interpolated property values with 48 property aliases set by a multi-value template variable (12 values)

image

More general tests covered:

1. Manually enter multiple property aliases in any query and get expected results
2. Manually enter property aliases and use one or more template variables (single or multi-value) in any query
3. Same as above for the propertyId field

**Which issue(s) this PR fixes**:

Fixes # https://github.com/grafana/iot-sitewise-datasource/issues/412

**Special notes for your reviewer**:

### CODE REVIEW

**Review by chejimmy — APPROVED** [review]
code looks good generally but I'm a lil hesitant to merge it
what do you say we do a code demo or bug bash before merge?

File: pkg/sitewise/api/util.go [hunk 1]
```diff
@@ -20,129 +20,184 @@ var (
...
 		return nil
 	}
-	return aws.String(query.PropertyAlias)
+	return aws.String(query.PropertyAliases[0])
 }
 
 func filterAnomalyAssetIds(ctx context.Context, client client.SitewiseClient, query models.AssetPropertyValueQuery) ([]string, error) {
 	anomalyAssetIds := []string{}
 
 	switch {
```
Comment (idastambuk) [inline pkg/sitewise/api/util.go:109]: Can we refactor this to an if/else statement, switch seems unnecessary

File: src/types.ts [hunk 2]
```diff
@@ -75,8 +75,14 @@ export interface SitewiseQuery extends DataQuery {
   assetId?: string;
   // One or more assets to filter -- when multiple, they should share the same properties, the batch API will be called
   assetIds?: string[];
+  /** @deprecated -- this is migrated to propertyIds */
   propertyId?: string;
+  // One or more properties to fetch data
+  propertyIds?: string[];
+  /** @deprecated -- this is migrated to propertyIds */
```
Comment (idastambuk) [inline src/types.ts:82]:
```suggestion
  /** @deprecated -- this is migrated to propertyAliases */
```

File: pkg/resource/sitewise.go [hunk 3]
```diff
@@ -29,13 +29,29 @@ func (rp *SitewiseResources) Asset(ctx context.Context, assetId string) (*iotsit
...
+		defaultOutput := &iotsitewise.DescribeAssetPropertyOutput{
 			AssetName: aws.String(""),
 			AssetProperty: &iotsitewise.Property{
 				Name:     aws.String(propertyAlias),
 				DataType: aws.String("?"),
 			},
-		}, nil
+		}
+
+		resp, err := rp.client.DescribeTimeSeriesWithContext(ctx, &iotsitewise.DescribeTimeSeriesInput{
```
[the remark this answers is not in the document]
Comment (hwandersman) [reply pkg/resource/sitewise.go:40]: We use the assetId/propertyId to describe the asset property and get its datatype, asset name, and other metadata to add to the resulting frame. This allows customers to see more information in the labels of the data result.

File: pkg/sitewise/api/util.go (hunk 1 above)
Comment (hwandersman) [reply pkg/sitewise/api/util.go:109]: Yes, I might have missed some places that had this two case switch statement

File: pkg/server/test/property_value_aggregate_test.go [hunk 4]
```diff
@@ -413,128 +381,263 @@ func TestPropertyValueAggregate_with_error(t *testing.T) {
...
-				Value:     &iotsitewise.Aggregates{Sum: Pointer(1688.6)},
-			}},
-			EntryId: aws.String(fmt.Sprintf("%dassetid-aaaa-2222-bbbb-3333cccc4444", i)),
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			mockSw := &mocks.SitewiseClient{}
+			mockedSuccessEntries := []*iotsitewise.BatchGetAssetPropertyAggregatesSuccessEntry{}
+			numBatch := 0
+
+			if tc.numPropertyAliases > 0 {
```
[the remark this answers is not in the document]
Comment (hwandersman) [reply pkg/server/test/property_value_aggregate_test.go:421]: Yes the conditional code could be moved to separate utils that builds the mocked responses for different test cases

File: src/components/query/visual-query-builder/PropertyQueryEditor.tsx [hunk 5]
```diff
@@ -133,30 +161,58 @@ export class PropertyQueryEditor extends PureComponent<Props, State> {
     onChange(newQuery);
   }
 
-  onPropertyChange = (sel: SelectableValue<string>) => {
-    const { onChange, query } = this.props;
-    const update = { ...query, propertyId: sel.value! };
+  onPropertyChange = (sel: SelectableValue<string> | Array<SelectableValue<string>>) => {
+    const propertyIds: Set<string> = new Set();
```
[the remark this answers is not in the document]
Comment (hwandersman) [reply src/components/query/visual-query-builder/PropertyQueryEditor.tsx:165]: You could technically enter the same propertyId manually, or have the same id resolved from template variables

File: src/components/query/visual-query-builder/PropertyQueryEditor.tsx [hunk 6]
```diff
@@ -253,40 +320,55 @@ export class PropertyQueryEditor extends PureComponent<Props, State> {
 
   render() {
     const { query, datasource, onChange } = this.props;
-    const { loading, assets, assetProperties } = this.state;
+    const { loading, assets, assetProperties, propertyAliases } = this.state;
+
+    let currentPropertyAlias = propertyAliases.filter((alias) =>
```
[the remark this answers is not in the document]
Comment (hwandersman) [reply src/components/query/visual-query-builder/PropertyQueryEditor.tsx:325]: Yeah I saw this with entering assetIds as "externalIds" too. Definitely needs to be fixed.

File: src/sitewiseCache.ts [hunk 7]
```diff
@@ -325,3 +316,16 @@ export const useModelsOptions = (cache: SitewiseCache): { isLoading: boolean; op
 
   return { isLoading, options };
 };
+export function getPropertyIdPickerOptions(asset?: AssetInfo): Array<SelectableValue<string>> {
+  const options: Array<SelectableValue<string>> = [];
+  if (asset?.properties) {
+    for (const property of asset.properties) {
```
[the remark this answers is not in the document]
Comment (diehbria) [reply src/sitewiseCache.ts]: resolved

**Comment by chejimmy** [comment]
Update (25/03/06): single property selection in UI until a better UX is designed around pairing assets and properties

File: src/types.ts (hunk 2 above)
Comment (chejimmy) [reply src/types.ts:79]: resolving in new commit update

File: src/components/query/visual-query-builder/PropertyQueryEditor.tsx [hunk 8]
```diff
@@ -86,22 +91,45 @@ export class PropertyQueryEditor extends PureComponent<Props, State> {
...
 
     const assetChanged = query?.assetIds !== oldProps?.query?.assetIds;
-    const propChanged = query?.propertyId !== oldProps?.query?.propertyId;
+    const propChanged = query?.propertyIds !== oldProps?.query?.propertyIds;
     const regionChanged = query?.region !== oldProps?.query?.region;
 
     if (assetChanged || propChanged || regionChanged) {
       if (!query.assetIds?.length && !regionChanged) {
-        this.setState({ assetId: undefined, asset: undefined, property: undefined, loading: false });
+        this.setState({ assetId: undefined, asset: undefined, loading: false });
```
[the remark this answers is not in the document]
Comment (chejimmy) [reply src/components/query/visual-query-builder/PropertyQueryEditor.tsx:100]: adding it 👍

File: src/components/query/visual-query-builder/PropertyQueryEditor.tsx [hunk 9]
```diff
@@ -123,7 +151,7 @@ export class PropertyQueryEditor extends PureComponent<Props, State> {
         ? {
             ...query,
             assetIds: [],
-            propertyId: undefined,
+            propertyIds: undefined,
```
[the remark this answers is not in the document]
Comment (chejimmy) [reply src/components/query/visual-query-builder/PropertyQueryEditor.tsx:155]: updating to array

File: src/components/query/visual-query-builder/PropertyQueryEditor.tsx [hunk 10]
```diff
@@ -133,30 +161,58 @@ export class PropertyQueryEditor extends PureComponent<Props, State> {
...
-    if (isAssetPropertyAggregatesQuery(update)) {
-      if (update.propertyId) {
-        const info = getAssetProperty(this.state.asset, update.propertyId);
+    const { onChange, query } = this.props;
 
-        if (!update.aggregates) {
-          update.aggregates = [];
-        }
+    const newQuery =
+      Array.isArray(sel) && sel.length === 0
```
[the remark this answers is not in the document]
Comment (chejimmy) [reply src/components/query/visual-query-builder/PropertyQueryEditor.tsx:180]: updating

File: src/components/query/visual-query-builder/PropertyQueryEditor.tsx [hunk 11]
```diff
@@ -133,30 +161,58 @@ export class PropertyQueryEditor extends PureComponent<Props, State> {
...
     }
 
-    onChange(update);
+    onChange(newQuery);
+  };
+
+  onSetPropertyAlias = (propertyAlias?: string) => {
+    const { onChange, query } = this.props;
+    if (!propertyAlias) {
+      onChange({ ...query, propertyAliases: undefined });
```
[the remark this answers is not in the document]
Comment (chejimmy) [reply src/components/query/visual-query-builder/PropertyQueryEditor.tsx:211]: updating

**Review by idastambuk — APPROVED** [review]
